File: src/driftbuster/core/detector.py
"""Core detection orchestration for DriftBuster.

Examples
--------
>>> from pathlib import Path
>>> from driftbuster.core.types import DetectionMatch
>>> from driftbuster.formats import format_registry as registry
>>> class StaticPlugin:
...     name = "static-fixture"
...     priority = 0
...     def detect(self, path, sample, text):
...         with path.open("rb") as handle:  # context manager best practice
...             _ = handle.read(1)
...         return DetectionMatch(
...             plugin_name=self.name,
...             format_name="fixture",
...             variant=None,
...             confidence=1.0,
...             reasons=["Static match for doctest"],
...         )
>>> result = scan_file(
...     Path(__file__), sample_size=64, plugins=(StaticPlugin(),)
... )
>>> result.format_name
'fixture'
>>> class PriorityPlugin:
...     name = "priority-fixture"
...     priority = 100
...     def detect(self, path, sample, text):
...         return DetectionMatch(
...             plugin_name=self.name,
...             format_name="fixture",
...             variant="ordered",
...             confidence=0.9,
...             reasons=["Manual ordering"],
...         )
>>> manual = Detector(
...     plugins=(PriorityPlugin(), StaticPlugin()),
...     sample_size=32,
...     sort_plugins=False,
... )
>>> manual.scan_file(Path(__file__)).plugin_name
'priority-fixture'
"""

from __future__ import annotations

import logging
from pathlib import Path
from typing import Callable, Iterable, List, Optional, Sequence

from ..catalog import DETECTION_CATALOG
from ..formats import format_registry as registry
from .profiles import ProfileStore, ProfiledDetection, normalize_tags
from .types import DetectionMatch, validate_detection_metadata

logger = logging.getLogger(__name__)

_DEFAULT_SAMPLE_SIZE = 128 * 1024  # 128 KiB default clamp.
_MAX_SAMPLE_SIZE = 512 * 1024  # Guardrail against excessive reads.

_DEFAULT_TOTAL_SAMPLE_BUDGET = 16 * 1024 * 1024  # 16 MiB aggregate guardrail.


class DetectorIOError(Exception):
    """Represents an I/O failure that occurred while scanning."""

    def __init__(self, path: Path, reason: str) -> None:
        self.path = Path(path)
        self.reason = reason
        super().__init__(f"{self.path}: {self.reason}")

    def __repr__(self) -> str:  # pragma: no cover - trivial wrapper
        path_repr = str(self.path)
        return f"DetectorIOError(path={path_repr!r}, reason={self.reason!r})"


def _wrap_io_error(path: Path, exc: OSError) -> DetectorIOError:
    """Convert an ``OSError`` into ``DetectorIOError``."""

    return DetectorIOError(path=Path(path), reason=str(exc))


def _validate_sample_size(sample_size: int) -> int:
    """Clamp and validate the requested sample size."""

    if sample_size <= 0:
        raise ValueError("sample_size must be a positive integer")
    if sample_size > _MAX_SAMPLE_SIZE:
        logger.warning(
            "Sample size %s exceeds %s bytes; clamping to guardrail.",
            sample_size,
            _MAX_SAMPLE_SIZE,
        )
        return _MAX_SAMPLE_SIZE
    return sample_size


def _validate_total_sample_budget(value: Optional[int]) -> Optional[int]:
    if value is None:
        return None
    if value <= 0:
        raise ValueError("max_total_sample_bytes must be a positive integer")
    return value


def _titleise_component(component: str) -> str:
    if not component:
        return component
    for index, char in enumerate(component):
        if char.isalpha():
            break
    else:
        return component
    prefix = component[:index]
    alpha = component[index].upper()
    suffix = component[index + 1:]
    return f"{prefix}{alpha}{suffix}"


def _normalise_reason_token(token: str) -> str:
    parts = token.split("-")
    normalised_parts = []
    for part in parts:
        subparts = part.split(":")
        normalised_subparts = [_titleise_component(sub) for sub in subparts]
        normalised_parts.append(":".join(normalised_subparts))
    return "-".join(normalised_parts)


def _normalise_reasons(reasons: Iterable[str]) -> List[str]:
    seen: set[str] = set()
    normalised: List[str] = []
    for raw in reasons:
        text = str(raw).strip()
        if not text:
            continue
        collapsed = " ".join(text.split())
        tokens = [
            _normalise_reason_token(token)
            for token in collapsed.split(" ")
        ]
        formatted = " ".join(tokens)
        if formatted not in seen:
            seen.add(formatted)
            normalised.append(formatted)
    return normalised


class Detector:
    """Coordinates registered format plugins to detect configuration types.

    Example
    -------
    >>> from pathlib import Path
    >>> from driftbuster.formats import format_registry as registry
    >>> class DummyPlugin(registry.FormatPlugin):
    ...     name = "dummy"
    ...     priority = 10
    ...     def detect(self, path, sample, text):
    ...         return None
    >>> detector = Detector(plugins=(DummyPlugin(),), sample_size=64)
    >>> detector.scan_file(Path(__file__)) is None
    True
    """

    def __init__(
        self,
        plugins: Optional[Sequence[registry.FormatPlugin]] = None,
        *,
        sample_size: Optional[int] = None,
        max_total_sample_bytes: Optional[int] = None,
        sort_plugins: bool = True,
        on_error: Optional[Callable[[Path, Exception], None]] = None,
    ) -> None:
        """Create a detector.

        Args:
            plugins: Optional explicit plugin sequence. When omitted the
                registry provided defaults are used.
            sample_size: Optional number of bytes to read from each file for
                detection heuristics. ``None`` uses ``_DEFAULT_SAMPLE_SIZE``.
            max_total_sample_bytes: Optional aggregate sampling budget. When
                ``None`` the default guardrail of ``_DEFAULT_TOTAL_SAMPLE_BUDGET``
                (16 MiB) is applied. Use a larger value when scanning very large
                directories.
            sort_plugins: Controls whether plugins are re-sorted by priority.
                Disable this when you inject a custom ordered sequence and want
                registration order to win over priority values.
            on_error: Optional callback invoked with ``(path, exception)`` when
                file system access fails. The exception passed will be a
                :class:`DetectorIOError`.
        """
        selected_plugins = list(plugins) if plugins is not None else list(
            registry.get_plugins()
        )
        if sort_plugins:
            selected_plugins = sorted(
                selected_plugins,
                key=lambda plugin: plugin.priority,
            )
        self._plugins = list(selected_plugins)
        validated_sample = _validate_sample_size(
            _DEFAULT_SAMPLE_SIZE if sample_size is None else sample_size
        )
        self._sample_size = validated_sample
        self._max_total_sample_bytes = _validate_total_sample_budget(
            max_total_sample_bytes
        )
        if self._max_total_sample_bytes is None:
            self._max_total_sample_bytes = _DEFAULT_TOTAL_SAMPLE_BUDGET
        self._consumed_sample_bytes = 0
        self._budget_exhausted = False
        self._on_error = on_error

    def _handle_error(
        self,
        path: Path,
        error: DetectorIOError,
        *,
        cause: Optional[BaseException] = None,
    ) -> None:
        if self._on_error is not None:
            try:
                self._on_error(path, error)
            except Exception:  # pragma: no cover - defensive
                logger.exception("on_error handler raised during scan")
        if cause is not None:
            raise error from cause
        raise error

    def reset_sample_budget(self) -> None:
        """Reset aggregate sampling counters for a fresh scan."""

        self._consumed_sample_bytes = 0
        self._budget_exhausted = False

    @property
    def sample_budget_exhausted(self) -> bool:
        return self._budget_exhausted

    @property
    def sample_budget_remaining(self) -> Optional[int]:
        if self._max_total_sample_bytes is None:
            return None
        remaining = self._max_total_sample_bytes - self._consumed_sample_bytes
        return max(0, remaining)

    def scan_file(self, path: Path) -> Optional[DetectionMatch]:
        path = Path(path)
        if not path.is_file():
            raise FileNotFoundError(f"Expected file path, got: {path}")

        if (
            self._max_total_sample_bytes is not None
            and self._consumed_sample_bytes >= self._max_total_sample_bytes
        ):
            self._budget_exhausted = True
            logger.warning(
                "Sample budget exhausted before scanning: %s", path
            )
            return None

        try:
            with path.open("rb") as handle:
                read_size = self._sample_size + 1
                remaining = None
                if self._max_total_sample_bytes is not None:
                    remaining = (
                        self._max_total_sample_bytes - self._consumed_sample_bytes
                    )
                    read_size = min(read_size, remaining + 1)
                raw = handle.read(read_size)
        except OSError as exc:
            self._handle_error(path, _wrap_io_error(path, exc), cause=exc)
            return None
        if (
            self._max_total_sample_bytes is not None
            and remaining is not None
        ):
            sample = raw[: min(self._sample_size, remaining)]
        else:
            sample = raw[: self._sample_size]
        truncated = len(raw) > self._sample_size
        text: Optional[str] = None
        encoding: Optional[str] = None
        if registry.looks_text(sample):
            text, encoding = registry.decode_text(sample)

        if self._max_total_sample_bytes is not None:
            self._consumed_sample_bytes += len(sample)
            if self._consumed_sample_bytes >= self._max_total_sample_bytes:
                self._budget_exhausted = True

        for plugin in self._plugins:
            match = plugin.detect(path, sample, text)
            if match is not None:
                if match.metadata is not None:
                    metadata = dict(match.metadata)
                else:
                    metadata = {}
                if "bytes_sampled" not in metadata:
                    metadata["bytes_sampled"] = len(sample)
                if encoding is not None and "encoding" not in metadata:
                    metadata["encoding"] = encoding
                    reason = f"Decoded content using {encoding} encoding"
                    if reason not in match.reasons:
                        match.reasons.append(reason)
                if truncated:
                    metadata["sample_truncated"] = True
                    reason = f"Truncated sample to {self._sample_size}B"
                    if reason not in match.reasons:
                        match.reasons.append(reason)
                if (
                    self._max_total_sample_bytes is not None
                    and self._budget_exhausted
                ):
                    metadata["sample_budget_exhausted"] = True
                    reason = (
                        f"Sampling budget exhausted after {len(sample)}B"
                    )
                    if reason not in match.reasons:
                        match.reasons.append(reason)
                match.metadata = metadata or None
                match.metadata = validate_detection_metadata(
                    match,
                    DETECTION_CATALOG,
                )
                match.reasons = _normalise_reasons(match.reasons)
                return match
        return None

    def scan_path(
        self,
        root: Path,
        glob: str = "**/*",
        *,
        reset_budget: bool = True,
    ) -> List[tuple[Path, Optional[DetectionMatch]]]:
        """Scan ``root`` while enforcing the aggregate sampling budget.

        Args:
            reset_budget: When ``True`` the detector resets its aggregate
                sampling counter before scanning. Set to ``False`` to continue
                consuming the existing budget across multiple directory roots.
        """
        root = Path(root)
        try:
            if root.is_file():
                if reset_budget:
                    self.reset_sample_budget()
                return [(root, self.scan_file(root))]

            if not root.is_dir():
                raise FileNotFoundError(f"Path does not exist: {root}")
        except OSError as exc:
            self._handle_error(root, _wrap_io_error(root, exc), cause=exc)
            return []

        if reset_budget:
            self.reset_sample_budget()
        results: List[tuple[Path, Optional[DetectionMatch]]] = []
        try:
            iterable = sorted(root.glob(glob))
        except OSError as exc:
            self._handle_error(root, _wrap_io_error(root, exc), cause=exc)
            return results
        for path in iterable:
            try:
                if path.is_file():
                    results.append((path, self.scan_file(path)))
                    if (
                        self._max_total_sample_bytes is not None
                        and self._budget_exhausted
                    ):
                        logger.warning(
                            "Sample budget exhausted while scanning %s;"
                            " skipping remaining paths under %s",
                            path,
                            root,
                        )
                        break
            except OSError as exc:
                self._handle_error(path, _wrap_io_error(path, exc), cause=exc)
        return results

    def scan_with_profiles(
        self,
        root: Path,
        *,
        profile_store: ProfileStore,
        tags: Optional[Sequence[str]] = None,
        glob: str = "**/*",
    ) -> List[ProfiledDetection]:
        """Scan ``root`` and annotate matches with configuration profiles.

        Args:
            root: File or directory to scan.
            profile_store: Registry providing configuration profiles.
            tags: Tag collection used to select relevant profiles/configs.
            glob: Optional glob pattern when scanning directories.
        """

        if profile_store is None:
            raise ValueError("profile_store must be provided")

        normalized_tags = normalize_tags(tags)
        root_path = Path(root)
        scan_results = self.scan_path(root_path, glob=glob)
        profiled: List[ProfiledDetection] = []

        root_is_dir = root_path.is_dir()

        for path, detection in scan_results:
            relative: Optional[str]
            if root_is_dir:
                try:
                    relative = path.relative_to(root_path).as_posix()
                except ValueError:
                    relative = path.name
            else:
                relative = path.name

            applied = profile_store.matching_configs(
                normalized_tags,
                relative_path=relative,
            )
            # If any matching config sets metadata flag to ignore review, annotate detection
            if detection and detection.metadata:
                try:
                    ignore = any(
                        bool(getattr(cfg.config, "metadata", {}).get("ignore_review_flags"))
                        for cfg in applied
                    )
                except Exception:
                    ignore = False
                if ignore and detection.metadata.get("needs_review"):
                    detection.metadata["review_ignored"] = True
                    detection.metadata["needs_review"] = False
            profiled.append(
                ProfiledDetection(
                    path=path,
                    detection=detection,
                    profiles=applied,
                )
            )

        return profiled


def scan_file(
    path: Path,
    *,
    sample_size: Optional[int] = None,
    plugins: Optional[Sequence[registry.FormatPlugin]] = None,
    sort_plugins: bool = True,
    on_error: Optional[Callable[[Path, Exception], None]] = None,
) -> Optional[DetectionMatch]:
    """Convenience wrapper that uses the default detector instance."""

    detector = Detector(
        plugins=plugins,
        sample_size=sample_size,
        sort_plugins=sort_plugins,
        on_error=on_error,
    )
    return detector.scan_file(Path(path))


def scan_path(
    root: Path,
    glob: str = "**/*",
    *,
    sample_size: Optional[int] = None,
    plugins: Optional[Sequence[registry.FormatPlugin]] = None,
    sort_plugins: bool = True,
    on_error: Optional[Callable[[Path, Exception], None]] = None,
) -> List[tuple[Path, Optional[DetectionMatch]]]:
    """Convenience wrapper mirroring :meth:`Detector.scan_path`."""

    detector = Detector(
        plugins=plugins,
        sample_size=sample_size,
        sort_plugins=sort_plugins,
        on_error=on_error,
    )
    return detector.scan_path(Path(root), glob=glob)

File: gui/DriftBuster.Backend/Detection/Detector.cs
using System.Diagnostics;
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

using DriftBuster.Backend.Detection.Catalog;
using DriftBuster.Backend.Infrastructure;
using DriftBuster.Backend.Profiles.Detection;

namespace DriftBuster.Backend.Detection;

/// <summary>Coordinates format plugins to detect configuration types under bounded sampling and an aggregate budget.</summary>
public class Detector
{
    /// <summary>Bytes read from each file when no sample size is requested.</summary>
    public const int DefaultSampleSize = 128 * 1024;

    /// <summary>Guardrail against excessive reads; larger requests are clamped with a warning.</summary>
    public const int MaxSampleSize = 512 * 1024;

    /// <summary>Aggregate sampling guardrail applied when no budget is requested.</summary>
    public const long DefaultTotalSampleBudget = 16L * 1024 * 1024;

    private readonly List<IFormatPlugin> _plugins;
    private readonly int _sampleSize;
    private readonly long _maxTotalSampleBytes;
    private readonly Action<string, Exception>? _onError;
    private readonly Action<string> _warn;
    private long _consumedSampleBytes;
    private bool _budgetExhausted;

    /// <param name="plugins">Explicit plugin sequence; null uses <see cref="DefaultPlugins"/>.</param>
    /// <param name="sampleSize">Bytes read from each file; null uses <see cref="DefaultSampleSize"/>.</param>
    /// <param name="maxTotalSampleBytes">Aggregate sampling budget; null uses <see cref="DefaultTotalSampleBudget"/>.</param>
    /// <param name="sortPlugins">Re-sort plugins by priority (stable); disable to keep the given order.</param>
    /// <param name="onError">Invoked with (path, <see cref="DetectorIOException"/>) before the error is raised.</param>
    /// <param name="onWarning">Receives guardrail warnings; defaults to <see cref="Trace.TraceWarning(string)"/>.</param>
    public Detector(
        IEnumerable<IFormatPlugin>? plugins = null,
        int? sampleSize = null,
        long? maxTotalSampleBytes = null,
        bool sortPlugins = true,
        Action<string, Exception>? onError = null,
        Action<string>? onWarning = null)
    {
        _warn = onWarning ?? (message => Trace.TraceWarning(message));
        var selected = plugins is null ? DefaultPlugins.GetPlugins().ToList() : plugins.ToList();
        if (sortPlugins)
        {
            selected = selected.OrderBy(plugin => plugin.Priority).ToList();
        }

        _plugins = selected;
        _sampleSize = ValidateSampleSize(sampleSize ?? DefaultSampleSize);
        _maxTotalSampleBytes = ValidateTotalSampleBudget(maxTotalSampleBytes) ?? DefaultTotalSampleBudget;
        _consumedSampleBytes = 0;
        _budgetExhausted = false;
        _onError = onError;
    }

    /// <summary>The plugins in the order they are consulted.</summary>
    public IReadOnlyList<IFormatPlugin> Plugins => _plugins;

    /// <summary>The effective (clamped) per-file sample size.</summary>
    public int SampleSize => _sampleSize;

    public bool SampleBudgetExhausted => _budgetExhausted;

    public long SampleBudgetRemaining => Math.Max(0, _maxTotalSampleBytes - _consumedSampleBytes);

    /// <summary>Resets the aggregate sampling counters for a fresh scan.</summary>
    public void ResetSampleBudget()
    {
        _consumedSampleBytes = 0;
        _budgetExhausted = false;
    }

    private int ValidateSampleSize(int sampleSize)
    {
        if (sampleSize <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(sampleSize), "sample_size must be a positive integer.");
        }

        if (sampleSize > MaxSampleSize)
        {
            _warn(string.Format(
                CultureInfo.InvariantCulture,
                "Sample size {0} exceeds {1} bytes; clamping to guardrail.",
                sampleSize,
                MaxSampleSize));
            return MaxSampleSize;
        }

        return sampleSize;
    }

    private static long? ValidateTotalSampleBudget(long? value)
    {
        if (value is null)
        {
            return null;
        }

        if (value <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(value), "max_total_sample_bytes must be a positive integer.");
        }

        return value;
    }

    /// <summary>
    /// Reports <paramref name="error"/> to the error callback and raises it. Subclasses may override to swallow
    /// errors, in which case the scan continues past the failing path.
    /// </summary>
    protected internal virtual void HandleError(string path, DetectorIOException error)
    {
        ArgumentNullException.ThrowIfNull(error);
        if (_onError is not null)
        {
            try
            {
                _onError(path, error);
            }
            catch (Exception exc) when (exc is not OutOfMemoryException)
            {
                _warn($"on_error handler raised during scan: {exc}");
            }
        }

        throw error;
    }

    /// <summary>Opens the file whose first bytes are sampled; overridable for fault injection.</summary>
    protected internal virtual Stream OpenFile(string path) => new FileStream(EnginePath.KernelPath(path), FileMode.Open, FileAccess.Read, FileShare.ReadWrite | FileShare.Delete);

    /// <summary><see cref="EnginePath.IsFile"/>; overridable for fault injection.</summary>
    protected internal virtual bool IsFile(string path) => EnginePath.IsFile(path);

    internal static string? ResolvePhysicalPath(string fullPath) => EnginePath.ResolvePhysicalPath(fullPath);

    /// <summary>Absolute paths (".." kept) of the sorted glob matches under <paramref name="root"/>; overridable for fault injection.</summary>
    protected internal virtual IReadOnlyList<string> EnumerateFiles(string root, string glob)
        => EnginePath.SortedGlob(root, glob).Select(EnginePath.Absolute).ToList();

    private byte[] ReadSample(string path, int readSize)
    {
        using var stream = OpenFile(path);
        var buffer = new byte[readSize];
        var total = 0;
        while (total < readSize)
        {
            var read = stream.Read(buffer, total, readSize - total);
            if (read == 0)
            {
                break;
            }

            total += read;
        }

        return total == readSize ? buffer : buffer[..total];
    }

    /// <summary>
    /// Samples <paramref name="path"/> and returns the first plugin match, enriched with sampling metadata, validated
    /// against the catalog and with normalised reasons; null when no plugin matched or the budget was already spent.
    /// </summary>
    public DetectionMatch? ScanFile(string path)
    {
        ArgumentNullException.ThrowIfNull(path);
        if (!IsFile(path))
        {
            throw new FileNotFoundException($"Expected file path, got: {path}", path);
        }

        if (_consumedSampleBytes >= _maxTotalSampleBytes)
        {
            _budgetExhausted = true;
            _warn($"Sample budget exhausted before scanning: {path}");
            return null;
        }

        var remaining = _maxTotalSampleBytes - _consumedSampleBytes;
        var readSize = (int)Math.Min(_sampleSize + 1L, remaining + 1);
        byte[] raw;
        try
        {
            raw = ReadSample(path, readSize);
        }
        catch (Exception exc) when (exc is IOException or UnauthorizedAccessException)
        {
            HandleError(path, new DetectorIOException(path, exc.Message, exc));
            return null;
        }

        var sampleLength = (int)Math.Min(Math.Min(_sampleSize, remaining), raw.Length);
        var sample = sampleLength == raw.Length ? raw : raw[..sampleLength];
        var truncated = raw.Length > _sampleSize;
        string? text = null;
        string? encoding = null;
        if (FormatRegistry.LooksText(sample))
        {
            (text, encoding) = FormatRegistry.DecodeText(sample);
        }

        _consumedSampleBytes += sample.Length;
        if (_consumedSampleBytes >= _maxTotalSampleBytes)
        {
            _budgetExhausted = true;
        }

        var first = FirstMatch(path, sample, text);
        return first is null ? null : Enrich(first, sample, encoding, truncated);
    }

    // First plugin match in registry order. A pattern that gives up on this sample must not abort the scan of every
    // other file, so a match timeout is reported the way an unreadable file is; this never
    // fires on a sample the engine matches in bounded time.
    private DetectionMatch? FirstMatch(string path, byte[] sample, string? text)
    {
        foreach (var plugin in _plugins)
        {
            DetectionMatch? match;
            try
            {
                match = plugin.Detect(path, sample, text);
            }
            catch (RegexMatchTimeoutException exc)
            {
                HandleError(path, new DetectorIOException(path, $"{plugin.Name} plugin timed out matching the sample", exc));
                return null;
            }

            if (match is not null)
            {
                return match;
            }
        }

        return null;
    }

    private DetectionMatch Enrich(DetectionMatch match, byte[] sample, string? encoding, bool truncated)
    {
        var metadata = match.Metadata;
        metadata.TryAdd("bytes_sampled", sample.Length);

        if (encoding is not null && !metadata.ContainsKey("encoding"))
        {
            metadata["encoding"] = encoding;
            AppendReason(match, $"Decoded content using {encoding} encoding");
        }

        if (truncated)
        {
            metadata["sample_truncated"] = true;
            AppendReason(match, string.Format(CultureInfo.InvariantCulture, "Truncated sample to {0}B", _sampleSize));
        }

        if (_budgetExhausted)
        {
            metadata["sample_budget_exhausted"] = true;
            AppendReason(match, string.Format(CultureInfo.InvariantCulture, "Sampling budget exhausted after {0}B", sample.Length));
        }

        DetectionMetadata.ValidateDetectionMetadata(match, DetectionCatalog.Default);
        match.Reasons = NormaliseReasons(match.Reasons);
        return match;
    }

    private static void AppendReason(DetectionMatch match, string reason)
    {
        if (!match.Reasons.Contains(reason))
        {
            match.Reasons.Add(reason);
        }
    }

    /// <summary>
    /// Scans a file or directory within the aggregate sampling budget. Only regular files are scanned; an undecodable name and
    /// a missing root are reported through <see cref="HandleError"/>. A walk stops after the file that exhausts the budget.
    /// </summary>
    /// <param name="resetBudget">Reset the aggregate counter first; false continues a budget across roots.</param>
    public virtual IReadOnlyList<(string Path, DetectionMatch? Match)> ScanPath(string root, string glob = "**/*", bool resetBudget = true)
    {
        ArgumentNullException.ThrowIfNull(root);
        // Normalised so "x.config/" names the file.
        root = LexicalPath.Str(root);
        var results = new List<(string Path, DetectionMatch? Match)>();
        try
        {
            if (IsFile(root))
            {
                if (resetBudget)
                {
                    ResetSampleBudget();
                }

                results.Add((root, ScanFile(root)));
                return results;
            }

            if (!Directory.Exists(EnginePath.KernelPath(root)))
            {
                throw new FileNotFoundException($"Path does not exist: {root}", root);
            }
        }
        catch (Exception exc) when (exc is IOException and not DetectorIOException || exc is UnauthorizedAccessException)
        {
            // A DetectorIOException raised by ScanFile has already been reported through HandleError; it propagates as is.
            HandleError(root, new DetectorIOException(root, exc.Message, exc));
            return results;
        }

        if (resetBudget)
        {
            ResetSampleBudget();
        }

        IReadOnlyList<string> candidates;
        try
        {
            candidates = EnumerateFiles(root, glob);
        }
        catch (Exception exc) when (exc is IOException or UnauthorizedAccessException)
        {
            HandleError(root, new DetectorIOException(root, exc.Message, exc));
            return results;
        }

        ScanCandidates(root, candidates, results);
        return results;
    }

    private void ScanCandidates(string root, IReadOnlyList<string> candidates, List<(string Path, DetectionMatch? Match)> results)
    {
        foreach (var path in candidates)
        {
            try
            {
                if (!IsFile(path))
                {
                    if (EnginePath.IsUndecodableName(path))
                    {
                        // The runtime cannot name the file, so the entry is reported.
                        HandleError(path, new DetectorIOException(path, "File name is not valid UTF-8; the entry cannot be opened"));
                    }

                    continue;
                }

                results.Add((path, ScanFile(path)));
                if (_budgetExhausted)
                {
                    _warn($"Sample budget exhausted while scanning {path}; skipping remaining paths under {root}");
                    break;
                }
            }
            catch (Exception exc) when (exc is IOException and not DetectorIOException || exc is UnauthorizedAccessException)
            {
                HandleError(path, new DetectorIOException(path, exc.Message, exc));
            }
        }
    }

    /// <summary>
    /// Scans <paramref name="root"/> and adds the profile configs that apply to each result (matched on the path relative to the
    /// root, or the file name). A config whose metadata sets <c>ignore_review_flags</c> clears <c>needs_review</c> and sets
    /// <c>review_ignored</c>; an error reading the configs counts as not ignoring.
    /// </summary>
    public IReadOnlyList<ProfiledDetection> ScanWithProfiles(
        string root,
        IProfileMatcher profileStore,
        IEnumerable<string?>? tags = null,
        string glob = "**/*")
    {
        ArgumentNullException.ThrowIfNull(root);
        if (profileStore is null)
        {
            throw new ArgumentNullException(nameof(profileStore), "profile_store must be provided.");
        }

        var normalizedTags = (tags ?? []).OfType<string>().Select(tag => tag.Trim()).Where(tag => tag.Length > 0).ToHashSet(StringComparer.Ordinal);
        var scanResults = ScanPath(root, glob);
        var profiled = new List<ProfiledDetection>();
        var rootIsDir = Directory.Exists(EnginePath.KernelPath(root));

        foreach (var (path, detection) in scanResults)
        {
            var relative = rootIsDir ? LexicalPath.RelativeTo(path, root) ?? PathText.Name(path) : PathText.Name(path);
            var applied = profileStore.MatchingConfigs(normalizedTags, relative);
            if (detection?.Metadata is { } metadata && applied.Any(cfg => cfg.Config.IgnoreReviewFlags))
            {
                if (metadata["needs_review"]?.GetValueKind() == JsonValueKind.True)
                {
                    metadata["review_ignored"] = true;
                    metadata["needs_review"] = false;
                }
            }

            profiled.Add(new ProfiledDetection(path, detection, applied));
        }

        return profiled;
    }

    /// <summary>Convenience wrapper for a single file: a fresh detector per call.</summary>
    public static DetectionMatch? ScanFileWithDefaults(
        string path,
        int? sampleSize = null,
        IEnumerable<IFormatPlugin>? plugins = null,
        bool sortPlugins = true,
        Action<string, Exception>? onError = null)
    {
        var detector = new Detector(plugins, sampleSize, sortPlugins: sortPlugins, onError: onError);
        return detector.ScanFile(path);
    }

    /// <summary>Convenience wrapper for a tree: a fresh detector per call.</summary>
    public static IReadOnlyList<(string Path, DetectionMatch? Match)> ScanPathWithDefaults(
        string root,
        string glob = "**/*",
        int? sampleSize = null,
        IEnumerable<IFormatPlugin>? plugins = null,
        bool sortPlugins = true,
        Action<string, Exception>? onError = null)
    {
        var detector = new Detector(plugins, sampleSize, sortPlugins: sortPlugins, onError: onError);
        return detector.ScanPath(root, glob);
    }

    // Upper-cases the first letter (any L* code point, astral included) with full case mapping (EngineText.Upper).
    private static string TitleiseComponent(string component)
    {
        if (component.Length == 0)
        {
            return component;
        }

        var offset = 0;
        foreach (var rune in component.EnumerateRunes())
        {
            if (EngineUnicode.IsAlpha(rune.Value))
            {
                return string.Concat(component.AsSpan(0, offset), EngineText.Upper(rune), component.AsSpan(offset + rune.Utf16SequenceLength));
            }

            offset += rune.Utf16SequenceLength;
        }

        return component;
    }

    private static string NormaliseReasonToken(string token)
    {
        var parts = token.Split('-');
        for (var index = 0; index < parts.Length; index++)
        {
            var subparts = parts[index].Split(':');
            for (var sub = 0; sub < subparts.Length; sub++)
            {
                subparts[sub] = TitleiseComponent(subparts[sub]);
            }

            parts[index] = string.Join(':', subparts);
        }

        return string.Join('-', parts);
    }

    /// <summary>
    /// Trims, collapses whitespace, title-cases the first letter of every '-' and ':' sub-part of each token and
    /// drops empty or duplicate reasons while preserving order.
    /// </summary>
    internal static IList<string> NormaliseReasons(IEnumerable<string> reasons)
    {
        var seen = new HashSet<string>(StringComparer.Ordinal);
        var normalised = new List<string>();
        foreach (var raw in reasons)
        {
            var text = EngineText.Strip(raw ?? string.Empty);
            if (text.Length == 0)
            {
                continue;
            }

            var words = EngineText.Split(text);
            var formatted = string.Join(' ', words.Select(NormaliseReasonToken));
            if (seen.Add(formatted))
            {
                normalised.Add(formatted);
            }
        }

        return normalised;
    }
}
